Reapply "nixos/dovecot: improve and harden systemd unit"

mweinelt opened the pull request:
This reverts commit a794031c597bd8a1898a1305b70952508e2d3612 , reapplying https://github.com/NixOS/nixpkgs/pull/418722.

With three new additions:

Allow `CAP_KILL`, so the dovecot master process may interrupt its child
processes.

Allow new privileges, so dovecot and call the setuid sendmail executable.

Allow AF_NETLINK sockets, so dovecot sieve handling can use the
getifaddrs syscall.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
Reapply "nixos/dovecot: improve and harden systemd unit"

This reverts commit a794031c597bd8a1898a1305b70952508e2d3612.

With the following additions:

Allow `CAP_KILL`, so the dovecot master process may interrupt its child
processes.

Allow new privileges, so dovecot and call the setuid sendmail executable.

Allow AF_NETLINK sockets, so dovecot sieve handling can use the
getifaddrs syscall.

Finally, we now asssert, that no options are set on the legacy dovecot2
systemd unit name, to make the user aware they need to update their
overrides.

exi commented:
Just a hint:
Renaming the systemd service also broke my config because i had overrides setup for `dovecot2.service`.
We should at least include an assertion or something on the old service name to make it fail for people still using it.

exi commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -692,23 +692,69 @@ in
     environment.etc."dovecot/dovecot.conf".source = cfg.configFile;
 
-    systemd.services.dovecot2 = {
+    systemd.services.dovecot = {
```
Renaming the systemd service also broke my config because i had overrides setup for dovecot2.service.
We should at least include an assertion or something on the old service name to make it fail for people still using it.

mweinelt replied:
Added an assertion when `config.systemd.services.dovecot2` is set.

exi replied:
Just out of curiosity, why even rename it at all?

2xsaiko replied:
This is the reason why I originally suggested 


```suggestion
    systemd.services.dovecot2 = {
      name = "dovecot";
```

btw; however ideally of course the attr and service name match up.

> Now I Just need to find a clean way in nixos-mailserver to understand whether dovecot or dovecot2 is the unit name used.

Add an internal option services.dovecot2.serviceNameIsDovecot, set it to true here and check whether it exists and is true in nixos-mailserver? Kinda awful but works (:

mweinelt replied:
Added in 4424fa674fc5

mweinelt commented on nixos/tests/dovecot.nix:
```diff
@@ -84,11 +84,13 @@
     machine.succeed("test-imap")
     machine.succeed("test-pop")
+
+    machine.log(machine.succeed("systemd-analyze security dovecot.service | grep -v ✓"))
```
This prints all the hardening options, and we only care if new ones appear, that are not positive yet.

The output of `systemd-analyze security` has the following format and I think the check is pretty clear once you've seen it.

```
✓ PrivateDevices=                                             Service has no access to hardware devices                                                                   
✗ ProtectHome=                                                Service has full access to home directories                                                              0.2
```

mweinelt committed:
nixos/dovecot: add marker to indicate new systemd unit name

This is a helper to escape infinite recursion when trying to check
whether the unit name is set before setting options on it.

mweinelt commented:
Here goes nothing.
